File: 33 HackerRank.go
// Go version of the python code
// 100%

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func insert(array []byte, count int, target byte) []byte {
	newArr := make([]byte, count+len(array))
	for k := 0; k < count; k++ {
		newArr[k] = target
	}
	copy(newArr[count:], array)
	return newArr
}

func nextPermutation(a []byte) bool {
	i := len(a) - 1
	for i > 0 && a[i-1] >= a[i] {
		i--
	}
	if i <= 0 {
		return false
	}

	j := len(a) - 1
	for a[j] <= a[i-1] {
		j--
	}

	a[i-1], a[j] = a[j], a[i-1]

	for j, k := i, len(a)-1; j < k; j, k = j+1, k-1 {
		a[j], a[k] = a[k], a[j]
	}
	return true
}

func numToCharArray(x, digits int) []byte {
	s := fmt.Sprintf("%0*d", digits, x)
	return []byte(s)
}

func merge(strFill, mask []byte) int {
	index := 0
	result := 0
	for _, m := range mask {
		result *= 10
		if m == '.' {
			result += int(strFill[index] - '0')
			index++
		} else {
			result += int(m - '0')
		}
	}
	return result
}

func clone(s []byte) []byte {
	cp := make([]byte, len(s))
	copy(cp, s)
	return cp
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanLines)
	scanner.Scan()
	inputLine := scanner.Text()
	parts := strings.Fields(inputLine)
	if len(parts) < 2 {
		fmt.Println("Please provide two integers N and K.")
		return
	}
	N, err1 := strconv.Atoi(parts[0])
	K, err2 := strconv.Atoi(parts[1])
	if err1 != nil || err2 != nil {
		fmt.Println("Invalid input.")
		return
	}
	keep := N - K

	// Tens array for base powers up to 10^4.
	Tens := []int{1, 10, 100, 1000, 10000}
	sumN := 0
	sumD := 0
	used := make(map[int]bool)

	for d := 1; d < Tens[keep]; d++ {
		for n := 1; n < d; n++ {
			charN := numToCharArray(n, keep)
			charD := numToCharArray(d, keep)
			for i := Tens[K-1]; i < Tens[K]; i++ {
				in := numToCharArray(i, K)
				isAscending := true
				for j := 1; j < len(in); j++ {
					if in[j-1] > in[j] {
						isAscending = false
						break
					}
				}
				if !isAscending {
					continue
				}
				pattern := insert(in, keep, '.')
				charInsertN := clone(pattern)
				for {
					newN := merge(charN, charInsertN)
					if newN >= Tens[N-1] {
						charInsertD := clone(pattern)
						for {
							newD := merge(charD, charInsertD)
							if newN*d == newD*n {
								id := newN*10000 + newD
								if !used[id] {
									sumN += newN
									sumD += newD
									used[id] = true
								}
							}
							if !nextPermutation(charInsertD) {
								break
							}
						}
					}
					if !nextPermutation(charInsertN) {
						break
					}
				}
			}
		}
	}
	fmt.Println(sumN, sumD)
}

File: 33 HackerRank.py
#  66.67% HackerRank
#  wrote same in C++ and then it got  83.33%
#  and then I wrote it in Go and it got 100%   -  (Go has very fast comilation)
#
#   Language	|    Compilation Time  |   Execution Time |       Notes
#   ---------------------------------------------------------------------------------
#   Python	    |    🚫 No compilation|   🐢 Slowest	 | Interpreted, great for quick prototyping, not for speed
#   C++	        |    🐌 Slow	      |   ⚡ Very Fast	| Powerful but long compile times, especially with templates
#   Go	        |    ⚡ Very Fast	 |   ⚡ Fast	       | One of Go’s biggest strengths — fast compiles and good performance

def insert(array, count, target):
    return [target] * count + array

def next_permutation(array):
    i = len(array) - 2
    while i >= 0 and array[i] >= array[i + 1]:
        i -= 1
    if i < 0:
        return False
    j = len(array) - 1
    while array[j] <= array[i]:
        j -= 1
    array[i], array[j] = array[j], array[i]
    array[i + 1:] = reversed(array[i + 1:])
    return True

def num_to_char_array(x, digits):
    result = str(x)
    while len(result) < digits:
        result = "0" + result
    return list(result)

def merge(str_fill, mask):
    index = 0
    result = 0
    for m in mask:
        result *= 10
        if m == '.':
            result += int(str_fill[index])
            index += 1
        else:
            result += int(m)
    return result

def solve():
    n_input, k_input = map(int, input().split())
    n = n_input
    k = k_input
    keep = n - k
    tens = [1, 10, 100, 1000, 10000]
    sum_n = 0
    sum_d = 0
    used = set()

    for d in range(1, tens[keep]):
        for num in range(1, d):
            char_n = num_to_char_array(num, keep)
            char_d = num_to_char_array(d, keep)
            for i in range(tens[k - 1], tens[k]):
                in_num = num_to_char_array(i, k)
                is_ascending = True
                for j in range(1, len(in_num)):
                    if in_num[j - 1] > in_num[j]:
                        is_ascending = False
                        break
                if is_ascending:
                    in_with_dots = insert(list(in_num), keep, '.')
                    char_insert_n = list(in_with_dots)
                    while True:
                        new_n = merge(char_n, char_insert_n)
                        if new_n >= tens[n - 1]:
                            char_insert_d = list(in_with_dots)
                            while True:
                                new_d = merge(char_d, char_insert_d)
                                if new_n * d == new_d * num:
                                    identifier = new_n * 10000 + new_d
                                    if identifier not in used:
                                        sum_n += new_n
                                        sum_d += new_d
                                        used.add(identifier)
                                if not next_permutation(char_insert_d):
                                    break
                        if not next_permutation(char_insert_n):
                            break

    print(f"{sum_n} {sum_d}")

if __name__ == "__main__":
    solve()